pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,448 @@
+		}
+	}
+	return nil, nil
+}
+
+func loadCertificate(certPath string) ([]*x509.Certificate, error) {
+	var certPEMBytes []byte
+	_, err := os.Stat(certPath)
+	if err == nil {
+		// Load the certificate from file.
+		certPEMBytes, err = ioutil.ReadFile(certPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read certificate file %s: %w", certPath, err)
+		}
+	} else if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to stat certificate file %s: %w", certPath, err)
+	}
+	if len(certPEMBytes) > 0 {
+		// Try to parse the certificate from the existing file.
+		certificates, err := certutil.ParseCertsPEM(certPEMBytes)
+		if err != nil {
+			klog.ErrorS(err, "Parse certificate from file error", "file", certPath)
+		} else {
+			return certificates, nil
+		}
+	}
+	return nil, nil
+}
+
+func (c *Controller) syncOVSConfigurations(certPath, keyPath, caPath string) error {
+	ovsConfig := map[string]interface{}{
+		ovsConfigCertificateKey:   certPath,
+		ovsConfigPrivateKeyKey:    keyPath,
+		ovsConfigCACertificateKey: caPath,
+	}
+	klog.InfoS("Updating OVS configurations for IPsec certificates")
+	return c.ovsBridgeClient.UpdateOVSOtherConfig(ovsConfig)
+}
+
+var newCSR = func(csrNamePrefix, commonName string, privateKey crypto.Signer) (*certificatesv1.CertificateSigningRequest, error) {
+	subject := &pkix.Name{
+		CommonName:   commonName,
+		Organization: []string{antreaapis.AntreaOrganizationName},
+	}
+	csrBytes, err := certutil.MakeCSR(privateKey, subject, []string{commonName}, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &certificatesv1.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: csrNamePrefix,
+		},
+		Spec: certificatesv1.CertificateSigningRequestSpec{
+			Request:    csrBytes,
+			SignerName: antreaapis.AntreaIPsecCSRSignerName,
+			Usages:     []certificatesv1.KeyUsage{certificatesv1.UsageIPsecTunnel},
+		},
+	}, nil
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting " + controllerName)
+	defer klog.InfoS("Shutting down " + controllerName)
+
+	// Load the previous configured certificate path from OVS database.
+	config, ovsErr := c.ovsBridgeClient.GetOVSOtherConfig()
+	if ovsErr != nil {
+		klog.ErrorS(ovsErr, "Failed to get OVS bridge other configs")
+	}
+
+	c.certPath = config[ovsConfigCertificateKey]
+	c.privateKeyPath = config[ovsConfigPrivateKeyKey]
+	if c.certPath != "" && c.privateKeyPath != "" {
+		pair, err := loadCertAndKeyFromFiles(c.caPath, c.certPath, c.privateKeyPath)
+		if err != nil {
+			klog.ErrorS(err, "Failed to load IPsec certificate and private key from existing files",
+				"ca", c.caPath, "cert", c.certPath, "key", c.privateKeyPath)
+		} else {
+			c.certificateKeyPair = pair
+		}
+	}
+
+	c.queue.Add(workerItemKey)
+	go wait.Until(c.worker, time.Second, stopCh)
+	<-stopCh
+}
+
+func (c *Controller) rotateCertificatesByCSR(nodeName string, certFolder string) (string, string, error) {
```

Comment:
nodeName and certFolder are member variables, shouldn't be needed here

Reply:
removed. thx!